rust/src/linalg/norm_l2.rs

```diff
@@ -12,6 +12,68 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use half::{bf16, f16};
+use num_traits::Float;
+
+/// L2 normalization
+pub trait Normalize<T: Float> {
+    type Output;
+
+    /// L2 Normalization over a Vector.
+    fn norm_l2(&self) -> Self::Output;
+}
+
+impl Normalize<f16> for &[f16] {
+    type Output = f16;
+
+    #[inline]
+    fn norm_l2(&self) -> Self::Output {
+        // Aarch64 does not have SIMD for f16
+        self.iter().map(|v| v * v).sum::<f16>().sqrt()
+    }
+}
+
+impl Normalize<bf16> for &[bf16] {
+    type Output = bf16;
+
+    #[inline]
+    fn norm_l2(&self) -> Self::Output {
+        // Aarch64 does not have SIMD for bf16
+        self.iter().map(|v| v * v).sum::<bf16>().sqrt()
+    }
+}
+
+impl Normalize<f32> for &[f32] {
+    type Output = f32;
+
+    #[inline]
+    fn norm_l2(&self) -> Self::Output {
+        #[cfg(target_arch = "aarch64")]
+        {
+            aarch64::neon::norm_l2(self)
+        }
+
+        #[cfg(target_arch = "x86_64")]
+        {
+            if is_x86_feature_detected!("fma") {
+                return x86_64::avx::norm_l2_f32(self);
+            }
+        }
+
+        #[cfg(not(target_arch = "aarch64"))]
+        self.iter().map(|v| v * v).sum::<f32>().sqrt()
+    }
+}
+
+impl Normalize<f64> for &[f64] {
```

Comment:
just curious if this is done for completeness reasons or if it was a user request.

Reply:
This is a placeholder for later to support f16/b16 index. For now, it is just for completeness.